llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -2666,6 +2667,46 @@ SDValue NVPTXTargetLowering::LowerShiftLeftParts(SDValue Op,
   }
 }
 
+// Lowers SETCC nodes that aren't directly supported by our arch.
```

Comment:
Should use setCondCodeAction and not reinvent all of this

Reply:
@arsenm Can you please explain a little further how I can use setCondCodeAction to avoid all of this?

Reply:
It's an additional legality table because the single type based legality rules are insufficient for compares. It's the same as setting expand / custom / promote, except for the specific condition code and type

Reply:
Hmm, this table seems to be used inconsistently, does the type represent the value which is produced by the SETCC node or the types of the values which are being compared?

https://github.com/llvm/llvm-project/blob/c1cec8c/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L8798
vs.
https://github.com/llvm/llvm-project/blob/c1cec8c/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L4994

Also what is the advantage of using this table over the current approach? What will I be able to avoid reinventing?

Reply:
The types being compared.

You are reinventing condition code legalization here: https://github.com/llvm/llvm-project/blob/c1cec8c0dc5b0296f0bc86745b867ff72c0a21e3/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L11832

And generic code can no longer accurately query which conditions are legal, which is why you have the hack to check if SETCC is legal

Reply:
Sorry, I'm still confused. I don't see anything like the boolean expansion I'm using for i1 in the LegalizeSetCCCondCode function. 

> And generic code can no longer accurately query which conditions are legal, which is why you have the hack to check if SETCC is legal

I think I'm going to need to add a check for legality to the dag combine no matter what, I suppose I can check this table instead. 

Also, is existing code like this incorrect? 

https://github.com/llvm/llvm-project/blob/c1cec8c0dc5b0296f0bc86745b867ff72c0a21e3/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L4994

Reply:
Okay, I've update the code to use the CondCode table. Not sure if this is what you had in mind though, I don't think it has made the implementation significantly simpler. I also did have to update a couple uses of the table which seemed incorrect. 

@arsenm does this look better?